lib/edge_impulse/ei_ncs.cpp

```diff
@@ -0,0 +1,399 @@
+		    size_t len)
+{
+	__ASSERT_NO_MSG((offset + len) <= INPUT_WINDOW_SIZE);
+
+	const float *buf_start = b->buf;
+	const float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	/* Processing head cannot change while processing is done. */
+	__ASSERT_NO_MSG(b->state == STATE_PROCESSING);
+
+	const float *read_start = b->processing_head + offset;
+	const float *read_end = read_start + len;
+
+	if ((read_end > buf_end) && (read_start < buf_end)) {
+		size_t buf_left = buf_end - read_start;
+
+		memcpy(b_res, read_start, buf_left * sizeof(b->buf[0]));
+		memcpy(b_res + buf_left, buf_start,
+		       (len - buf_left) * sizeof(b->buf[0]));
+	} else {
+		if (read_start >= buf_end) {
+			read_start -= ARRAY_SIZE(b->buf);
+		}
+		memcpy(b_res, read_start, len * sizeof(b->buf[0]));
+	}
+}
+
+static int buf_processing_move(struct data_buffer *b, size_t move,
+			       bool *process_buf)
+{
+	float *buf_start = b->buf;
+	float *buf_end = buf_start + ARRAY_SIZE(b->buf);
+
+	*process_buf = false;
+
+	k_spinlock_key_t key = k_spin_lock(&b->lock);
+
+	if (b->state == STATE_READY) {
+		b->state = STATE_PROCESSING;
+	} else {
+		__ASSERT_NO_MSG(b->state != STATE_DISABLED);
+		k_spin_unlock(&b->lock, key);
+		return -EBUSY;
+	}
+
+	size_t max_move = buf_calc_max_processing_move(b);
+
+	b->processing_head += move;
+	if (b->processing_head >= buf_end) {
+		b->processing_head -= ARRAY_SIZE(b->buf);
+	}
+
+	size_t processing_end_move = move + INPUT_WINDOW_SIZE;
+
+	if (processing_end_move > max_move) {
+		b->wait_data_size = processing_end_move - max_move;
+	} else {
+		*process_buf = true;
+	}
+
+	k_spin_unlock(&b->lock, key);
+
+	return 0;
+}
+
+bool ei_ncs_classifier_has_anomaly(void)
+{
+	return (HAS_ANOMALY) ? (true) : (false);
+}
+
+size_t ei_ncs_get_frame_size(void)
+{
+	return INPUT_FRAME_SIZE;
+}
+
+size_t ei_ncs_get_window_size(void)
+{
+	return INPUT_WINDOW_SIZE;
+}
+
+int ei_ncs_add_data(const float *data, size_t data_size)
+{
+	if (data_size % INPUT_FRAME_SIZE) {
+		return -EINVAL;
+	}
+
+	bool process_buf;
+	int res = buf_append(&ei_input, data, data_size, &process_buf);
+
+	if (process_buf) {
+		k_sem_give(&ei_sem);
+	}
+
+	return res;
+}
+
+int ei_ncs_clear_data(void)
+{
+	return buf_cleanup(&ei_input);
+}
+
+int ei_ncs_start_prediction(size_t window_shift, size_t frame_shift)
+{
+	size_t sample_shift = window_shift * ei_ncs_get_window_size() +
+			      frame_shift * ei_ncs_get_frame_size();
+
+	bool process_buf;
+	int ret = buf_processing_move(&ei_input, sample_shift, &process_buf);
+
+	if (process_buf) {
+		k_sem_give(&ei_sem);
+	}
+
+	return ret;
+}
+
+static int raw_feature_get_data(size_t offset, size_t length, float *out_ptr)
+{
+	buf_get(&ei_input, out_ptr, offset, length);
+
+	return 0;
+}
+
+static void processing_finished(int err)
+{
+	__ASSERT_NO_MSG(user_cb);
+
+	buf_processing_end(&ei_input);
+	user_cb(err);
+}
+
+static void edge_impulse_thread_fn(void)
+{
+	signal_t features_signal = {
```

Comment:
Maybe we should still put it on flash and cast away the const during the call? Note that here you depend on `run_classifier` on not changing `features_signal` even though the api does not say it is immutable.

Reply:
run_classifier is in general an external function that is part of the downloaded zip file. I would not assume that it will not modify features_signal unless it uses const for it.

Reply:
The behavior is not guaranteed and it may change in the future too.

Reply:
if so features_signal  should be always restored before the call. At the moment you set it only once.

Reply:
Ok, fixed